src/datasets/arrow_dataset.py

```diff
@@ -331,66 +382,108 @@ def to_tf_dataset(
-            dummy_labels (:obj:`bool`, default ``False``): If no `label_cols` are set, output an array of "dummy" labels
-                with each batch. This can avoid problems with `fit()` or `train_on_batch()` that expect labels to be
-                a Tensor or np.ndarray, but should (hopefully) not be necessary with our standard train_step().
+            drop_columns (:obj:`List[str]` or :obj:`str` or :obj:`bool`, optional): Dataset column(s) to drop before
+                passing to the `collate_fn`. By default, no columns are dropped.
             prefetch (:obj:`bool`, default ``True``): Whether to run the dataloader in a separate thread and maintain
                 a small buffer of batches for training. Improves performance by allowing data to be loaded in the
                 background while the model is training.
 
         Returns:
             :class:`tf.data.Dataset`
         """
-
+        # region Pre-checks and warnings
         if config.TF_AVAILABLE:
             import tensorflow as tf
         else:
             raise ImportError("Called a Tensorflow-specific function but Tensorflow is not installed.")
 
+        if isinstance(tf.distribute.get_strategy(), tf.distribute.TPUStrategy):
+            logger.warning(
+                "Note that to_tf_dataset() loads the data with a generator rather than a full tf.data "
+                "pipeline and is not compatible with remote TPU connections. If you encounter errors, please "
+                "try using a TPU VM or, if your data can fit in memory, loading it into memory as a dict of "
+                "Tensors instead of streaming with to_tf_dataset()."
+            )
+        # endregion
+
+        # region Argument and default value handling
+
+        # Our assumption at the end of argument processing is that either
         if collate_fn_args is None:
             collate_fn_args = {}
-
         if label_cols is None:
             label_cols = []
         elif isinstance(label_cols, str):
             label_cols = [label_cols]
         elif len(set(label_cols)) < len(label_cols):
             raise ValueError("List of label_cols contains duplicates.")
+        autodetected_columns = False
         if not columns:
-            raise ValueError("Need to specify at least one column.")
+            if not model:
+                raise ValueError("Either columns or model must be specified.")
+            autodetected_columns = True
+            columns = list(signature(model.call).parameters.keys())
+            model_name = type(model).__name__
+            # This is the same code as Sylvain's find_labels function in Transformers, but I didn't want the dependency
+            if not label_cols:
+                if "QuestionAnswering" in model_name:
+                    label_cols = [p for p in columns if "label" in p or p in ("start_positions", "end_positions")]
+                else:
+                    label_cols = [p for p in columns if "label" in p]
```

Comment:
This is very specific to Transformers models here, so you should check if the model is an instance of `TFPreTrainedModel` before doing the guessing.

Reply:
That's a very good point, yes! It's tricky, though, because these things are handled by `Trainer` for PyTorch users, where we have a lot of scope to add custom code, but if we want to use Keras for TF then these things have to be handled at the point of dataset creation. Maybe we could move this from a method on `Dataset` objects to a function in `transformers`, like `convert_to_tf_dataset()` that takes a HF `Dataset` as an argument?

Reply:
See my comment above.

Reply:
Ah sorry, yes! Should have paid more attention to the second paragraph.

Reply:
No worries ;-)

Reply:
Fixed!